### PR TITLE
PageStorage: remove wal files according to its max sequence when dump snapshot

### PR DESCRIPTION
**Pull request by lidezhu**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/7780

Problem Summary: After dump snapshot from memory for ps, it will remove old wal files. But new page data is always write into wal before applying to memory. So the data in memory may be less than the data in wal files. So using the snapshot dumped from memory to replace old wal files may cause some data lost.

### What is changed and how it works?
Add a cache to memorize the max sequence of wal files. 
When dump snapshot, traverse from the latest wal files and find the one whose max sequence is smaller than the snapshot's sequence. And remove the wal file and older ones.
We avoid maintaining each wal file's max sequence when writing, because it will make write process acquire one more lock which may cause performance penalty.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/WALStore.cpp [hunk 1]
```diff
@@ -270,4 +278,48 @@ bool WALStore::saveSnapshot(
...
+}
+
+UInt64 WALStore::getLogFileMaxSequence(const LogFilename & log_filename, std::function<UInt64(const String & record)> max_sequence_getter)
+{
+    std::unique_lock<std::mutex> lock(log_file_max_sequences_cache_mutex);
+    auto iter = log_file_max_sequences_cache.find(log_filename);
+    if (iter != log_file_max_sequences_cache.end())
+        return iter->second;
+    else
+    {
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/WALStore.cpp:304]: As we directly return in the previous `if (...)` statement, we can remove this `else` indent layer to be 

```
if (...)
    return ...;

...
```

File: dbms/src/Storages/Page/V3/WALStore.cpp [hunk 2]
```diff
@@ -270,4 +278,48 @@ bool WALStore::saveSnapshot(
...
+        while (true)
+        {
+            auto [ok, record] = log_reader->readRecord();
+            if (!ok)
+                break;
+
+            last_record = std::move(record);
+        }
+        if (last_record.empty())
+            return 0; // empty log file
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/WALStore.cpp:316]: ditto

File: dbms/src/Storages/Page/V3/WALStore.cpp [hunk 3]
```diff
@@ -270,4 +278,48 @@ bool WALStore::saveSnapshot(
...
+
+UInt64 WALStore::getLogFileMaxSequence(const LogFilename & log_filename, std::function<UInt64(const String & record)> max_sequence_getter)
+{
+    std::unique_lock<std::mutex> lock(log_file_max_sequences_cache_mutex);
+    auto iter = log_file_max_sequences_cache.find(log_filename);
+    if (iter != log_file_max_sequences_cache.end())
+        return iter->second;
+    else
+    {
+        auto log_reader = wal_store_reader->createLogReader(log_filename);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/WALStore.cpp:305]: Can we make `WALStoreReader::createLogReader` to be a static method? It is weird to me that we need to keep a `wal_store_reader` as long as `WALStore`

File: dbms/src/Storages/Page/V3/PageDirectory.cpp [hunk 4]
```diff
@@ -1825,12 +1833,9 @@ bool PageDirectory<Trait>::tryDumpSnapshot(const WriteLimiterPtr & write_limiter
     // The main reason write affect dumping snapshot is that we can not get a read-only
     // `being_ref_count` by the function `createSnapshot()`.
     assert(!files_snap.persisted_log_files.empty()); // should not be empty
-    auto log_num = files_snap.persisted_log_files.rbegin()->log_num;
-    auto identifier = fmt::format("{}.dump_{}", wal->name(), log_num);
 
     Stopwatch watch;
     // The records persisted in `files_snap` is older than or equal to all records in `edit`
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/PageDirectory.cpp]: Fixed

File: dbms/src/Storages/Page/V3/WALStore.cpp (hunk 1 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/WALStore.cpp]: Fixed

File: dbms/src/Storages/Page/V3/WALStore.cpp (hunk 3 above)
Comment (lidezhu) [reply dbms/src/Storages/Page/V3/WALStore.cpp]: Fixed.

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/Storages/Page/V3/PageDirectory.h [hunk 5]
```diff
@@ -152,9 +152,16 @@ class MultiVersionRefCount
...
             {
-                versioned_ref_counts = nullptr;
+                if (new_versioned_ref_counts->empty())
+                {
+                    versioned_ref_counts = nullptr;
+                }
+                else
+                {
+                    versioned_ref_counts.swap(new_versioned_ref_counts);
+                }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageDirectory.h:164]:
```suggestion
                {
                    // There could be some new ref count created after `snap_seq`, we need to
                    // keep the newly added ref counts
                    versioned_ref_counts.swap(new_versioned_ref_counts);
                }
```

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**